src/Rings/msolve/f4.jl

```diff
@@ -0,0 +1,121 @@
+import msolve_jll: libneogb
+import Libdl: dlopen, dlsym, dlclose
+
+export f4
+
+@doc Markdown.doc"""
+    f4(I, <keyword arguments>)
+
+Compute a Groebner basis of the given ideal I w.r.t. to the degree reverse lexicographical monomial ordering using Faugère's F4 algorithm.
+See [Fau99](@cite) for more information.
+
+**Note**: At the moment only ground fields of characteristic `p`, `p` prime, `p < 2^{31}` are supported.
+
+# Arguments
+- `Ì::MPolyIdeal`: input ideal.
+- `initial_hts::Int=17`: initial hash table size `log_2`.
+- `nr_thrds::Int=1`: number of threads for parallel linear algebra.
+- `max_nr_pairs::Int=0`: maximal number of pairs per matrix, only bounded by minimal degree if `0`.
+- `la_option::Int=2`: linear algebra option: exact sparse-dense (`1`), exact sparse (`2`, default), probabilistic sparse-dense (`42`), probabilistic sparse(`44`).
+- `reduce_gb::Int=1`: compute a reduced Gröbner basis for `I`
+- `info_level::Int=0`: info level printout: off (`0`, default), summary (`1`), detailed (`2`).
+
+# Examples
+```jldoctest
+julia> R,(x,y,z) = PolynomialRing(FiniteField(101), ["x","y","z"])
+(Multivariate Polynomial Ring in x, y, z over Galois field with characteristic 101, gfp_mpoly[x, y, z])
+
+julia> I = ideal(R, [x+2*y+2*z-1, x^2+2*y^2+2*z^2-x, 2*x*y+2*y*z-y])
+ideal(x + 2*y + 2*z + 100, x^2 + 100*x + 2*y^2 + 2*z^2, 2*x*y + 2*y*z + 100*y)
+
+julia> f4(I)
+Oscar.BiPolyArray{gfp_mpoly}(gfp_mpoly[#undef, #undef, #undef, #undef], Singular ideal over Singular Polynomial Ring (ZZ/101),(x,y,z),(dp(3),C) with generators (x + 2*y + 2*z - 1, y*z - 19*z^2 + 10*y + 40*z, y^2 - 41*z^2 + 20*y - 20*z, z^3 + 28*z^2 - 37*y + 13*z), Multivariate Polynomial Ring in x, y, z over Galois field with characteristic 101, Singular Polynomial Ring (ZZ/101),(x,y,z),(dp(3),C), true, #undef)
+```
+"""
+function f4(
+        I::MPolyIdeal;
+        initial_hts::Int=17,
+        nr_thrds::Int=1,
+        max_nr_pairs::Int=0,
+        la_option::Int=2,
+        reduce_gb::Int=1,
+        info_level::Int=0
+        )
+
+    singular_assure(I)
+    SI    = I.gens.S
+    R     = I.gens.Sx
+    # skip zero generators in ideal
+    ptr = Singular.libSingular.id_Copy(SI.ptr, R.ptr)
+    J   = Singular.Ideal(R, ptr)
+    Singular.libSingular.idSkipZeroes(J.ptr)
+    # get number of variables
+    nr_vars = Singular.nvars(R)
+    nr_gens = Singular.ngens(J)
+
+    mon_order   = 0
+    field_char  = Singular.characteristic(R)
+
+    # convert Singular ideal to flattened arrays of ints
+    if isprime(field_char)
+      lens, cfs, exps = convert_ff_singular_ideal_to_array(J, nr_vars, nr_gens)
+    else
+        error("At the moment f4 only supports finite fields.")
+    end
+    #= dir = joinpath(dirname(pathof(GroebnerBasis)),"../deps")
+     = lib = Libdl.dlopen("$dir/libmsolve.so.0.2.0") =#
+    lib = Libdl.dlopen(libneogb)
+    sym = Libdl.dlsym(lib, :f4_julia)
+
+    gb_ld   = ccall(:malloc, Ptr{Cint}, (Csize_t, ), sizeof(Cint))
```

Comment:
Can't this be done using julia? `Ref(Array{bla...})` or similar? The problem with a direct call to `malloc` is that Julia then does not know that there is memory involved. Ie. the `gc` does not know about this....

Reply:
But the memory is freed in exactly the same way some lines below. What would be the benefit of using Julia's gc handling this specific memory allocations?

Reply:
If you just need an `int &`, the julia constructor `Ref(Cint())` should do the job. I think the situation in https://github.com/Nemocas/Nemo.jl/pull/117 (see first review) is similar to this.

Reply:
Here are my comments after reviewing in more detail:
So, it looks like the f4_function has the following 4 troublesome arguments for output:
```
  C/C++         |  julia           | desc
----------------+------------------+--------------
   int& gb_ld   |  Ptr{Cint}       | number of polys in the output
  int*& gb_len  |  Ptr{Ptr{Cint}}  | array of lengths lengths of each poly in the output
  int*& gb_exp  |  Ptr{Ptr{Cint}}  | all output exponents (flattened)
  int*& gb_cf   |  Ptr{Ptr{Cvoid}} | all output coefficient (flattened)
```
The problem is of course if the `f4_julia` call throws. Here I am assuming that if
the `f4_julia` throws, then it leaves `gb_len`, `gb_exp`, and `gb_cf` undefined
(or aleast in a state that requires no cleanup). Then, on the julia side, I would
do something like the following. This avoids the annoying 4 small calls to malloc,
and also does not need any new c struct on the julia side to transport variables.
```julia
    gb_ld  = Ref(Cint(0))
    gb_len = Ref(Ptr{Cint}(0))
    gb_exp = Ref(Ptr{Cint}(0))
    gb_cf  = Ref(Ptr{Cvoid}(0))

... call f4_julia ...

    jl_ld   = gb_ld[]
    jl_len  = Base.unsafe_wrap(Array, gb_len[], jl_ld)
    jl_exp  = Base.unsafe_wrap(Array, gb_exp[], nr_terms*nr_vars)
    if 0 == char
        ptr   = reinterpret(Ptr{BigInt}, gb_cf[])
        jl_cf = Base.unsafe_wrap(Array, ptr, nr_terms)
    elseif Nemo.isprime(Nemo.FlintZZ(char))
        ptr   = reinterpret(Ptr{Int32}, gb_cf[])
        jl_cf = Base.unsafe_wrap(Array, ptr, nr_terms)
    end
```

Reply:
However, there is now the question of what `ccall((:free_julia_data, libneogb), ...)` does. 
1. Does `free_julia_data` free the (large amount of) array data representing the return polynomials? It seems the answer is "yes" here and if so, the comment `# convert to julia array, also give memory management to julia` is unclear as `;own=false` is the default in `unsafe_wrap`.
2. Does `free_julia_data` also free the 3 pointers `gb_len`, `gb_exp`, `gb_cf`  that were manually `malloc`d in the julia code? It would seem that answer here is "yes" otherwise the code here would be leaking, but then this means  that the approach of using julia `Ref`s for the output locations is not possible. :(

Reply:
What's about removing the `free_julia_data` call and doing all `unsafe_wrap` calls with `own=true`?

Reply:
I'm now doing the references stuff by @tthsqe12 combined with calling `msolve` and `f4` with another parameter being `jl_malloc`. I then use inside `msolve` `jl_malloc`to allocate the memory for the corresponding returned data. So julia should take care (gc) this data, right?

Is this a valid solution?

Reply:
I see you have removed the `free_julia_data` in the latest version and replaced it with my suggestion. Now the pointers `gb_len[]`, `gb_exp[]`, and `gb_cf[]` are not being cleaned up. Could you let us know exactly what `free_julia_data` was doing? Here is what I see:
1. `gb_len[]` This is just a pointer to bits types, so call whatever `free` on it to match the `malloc` used inside `f4_julia`.
2. `gb_exp[]` Ditto because of the bits type.
3. `gb_cf[]` in the `char > 0` the `void*` is another pointer to a bits type, so the interesting case is when `char = 0` and the `void*` is a `mpz_struct *`. In this case I have no idea how `Base.unsafe_wrap` behaves because the documentation for it is so scant. In order to cleanup of the array, we have to call `mpz_clear` on each entry and then `free` the pointer `gb_cf[]`, and the documentation for `unsafe_wrap` with `own=true` says it just frees the final pointer without cleaning up the entries. I hope someone more knowledgable in `Base.unsafe_wrap` can comment here.

Ideally, if a new free_julia_data was constructed something like this (this can also be written in julia)
```C
void new_free_julia_data(int* gb_len, int* gb_exp, void* gb_cf, int char, size_t nr_terms)
{
    free(gb_len);
    free(gb_exp);
    if (char > 0)
    {
        __mpz_struct * p = (__mpz_struct *) gb_cf;
        for (size_t i = 0; i < nr_terms; i++)
            mpz_clear(p + i);
    }
    free(gb_cf);
}
```
then the julia code looks like something
```julia
    gb_ld  = Ref(Cint(0))
    gb_len = Ref(Ptr{Cint}(0))
    gb_exp = Ref(Ptr{Cint}(0))
    gb_cf  = Ref(Ptr{Cvoid}(0))

... call f4_julia ...
... read data from gb_ld[], gb_len[], gb_exp[], gb_cf[] ...
... construct basis ...

    ccall((:new_free_julia_data, libneogb), Nothing,
          (Ptr{Cint}, Ptr{Cint}, Ptr{Cvoid}, Cint, Int),
          gb_len[], gb_exp[], gb_cf[], char, nr_terms)
```

With this approach, as long as `f4_julia` doesn't leave pointers that need to be cleaned up in `gb_len[], gb_exp[], gb_cf[]` when it is interrupted, we could say that it doesn't leak when interrupted (because the memory management of the references `gb_ld`, `gb_len` is julia's responsibility.)

Reply:
> I'm now doing the references stuff by @tthsqe12 combined with calling `msolve` and `f4` with another parameter being `jl_malloc`. I then use inside `msolve` `jl_malloc`to allocate the memory for the corresponding returned data. So julia should take care (gc) this data, right?
> 
> Is this a valid solution?

Is allocating an array of bits type with `jl_malloc` and then passing it to `Base.unsafe_wrap(ptr, len, own=true)` sufficient @thofma or @fingolfin? I have detailed above why I think this will be a problem in the characteristic 0 case, but lets get the char > 0 working first.

Reply:
(We discussed this on gather.town and came to a conclusion.)